pkg/util/kubernetes/kubelet/kubelet.go

```diff
@@ -497,44 +499,289 @@ func (ku *KubeUtil) setupKubeletApiEndpoint() error {
+	ku.kubeletApiEndpoint = fmt.Sprintf("https://%s:%d", ku.kubeletHost, httpsPort)
+	response, err := c.Get(ku.kubeletApiEndpoint + "/")
+	if err == nil {
+		log.Infof("Successfully queried to %s without any security settings, adding security transport settings to query %s%s", response.Request.URL, ku.kubeletApiEndpoint, kubeletPodPath)
+
+		response, err := ku.doKubeletRequest(kubeletPodPath)
+		if err == nil {
+			log.Infof("Successfully connected securely on kubelet endpoint %s", response.Request.URL)
+			switch {
+			case response.StatusCode == http.StatusOK:
+				log.Infof("Successfully authorized to query the kubelet on %s: 200, using %s as kubelet endpoint", response.Request.URL, ku.kubeletApiEndpoint)
+				ku.resetCredentials()
+				return nil
 
-	// HTTPS first
-	if _, errHTTPS := c.Get(fmt.Sprintf("https://%s:%d/", ku.kubeletHost, config.Datadog.GetInt("kubernetes_https_kubelet_port"))); errHTTPS != nil {
-		log.Debugf("Cannot connect through HTTPS: %s, trying through http", errHTTPS)
+			case response.StatusCode >= http.StatusInternalServerError:
+				log.Infof("Unexpected return code on request %s on kubelet endpoint %s", response.Request.URL, ku.kubeletApiEndpoint)
 
-		// Only try the HTTP if HTTPS failed
-		if _, errHTTP := c.Get(fmt.Sprintf("http://%s:%d/", ku.kubeletHost, config.Datadog.GetInt("kubernetes_http_kubelet_port"))); errHTTP != nil {
-			log.Debugf("Cannot connect through HTTP: %s", errHTTP)
-			return fmt.Errorf("cannot connect: https: %q, http: %q", errHTTPS, errHTTP)
+			case response.StatusCode == http.StatusUnauthorized:
+				log.Debugf("Unauthorized to request %s on kubelet endpoint %s, check the kubelet authentication/authorization settings", response.Request.URL, ku.kubeletApiEndpoint)
+
+			default:
+				log.Debugf("Unexpected http code %d on kubelet endpoint %s", response.StatusCode, ku.kubeletApiEndpoint)
+			}
+
+			// err != nil
+		} else if strings.Contains(err.Error(), "x509: certificate is valid for") {
+			log.Debugf(`Invalid x509 settings, the kubelet server certificate is not valid for this subject alternative name: %s, %v, Please check the SAN of the kubelet server certificate with "openssl x509 -in ${KUBELET_CERTIFICATE} -text -noout". `, ku.kubeletHost, err)
+			return err
+
+		} else if strings.Contains(err.Error(), "x509: certificate signed by unknown authority") {
+			log.Debugf(`The kubelet server certificate is signed by unknown authority, the current cacert is %s. Is the kubelet issuing self-signed certificates? Please validate the kubelet certificate with "openssl verify -CAfile %s ${KUBELET_CERTIFICATE}" to avoid this error: %v`, ku.rawConnectionInfo["ca_cert"], ku.rawConnectionInfo["ca_cert"], err)
+			return err
+
+		} else {
+			log.Debugf("Cannot query %s on kubelet endpoint %s: %v", kubeletPodPath, ku.kubeletApiEndpoint, err)
+			return err
 		}
+	} else {
+		log.Debugf("Cannot use the HTTPS endpoint: %s, trying through HTTP", ku.kubeletApiEndpoint)
```

Comment:
nitpick: since that function doesn't try with http, let's remove the `trying through http` part. The other function can print it if needed, but if we have it here and later decide to remove the http part, it will be misleading.